cmd/boulder-sa/main.go

```diff
@@ -1,41 +1,72 @@
 package main
 
 import (
+	"flag"
+	"os"
+
 	"github.com/jmhodges/clock"
 
 	"github.com/letsencrypt/boulder/cmd"
-	blog "github.com/letsencrypt/boulder/log"
 	"github.com/letsencrypt/boulder/metrics"
 	"github.com/letsencrypt/boulder/rpc"
 	"github.com/letsencrypt/boulder/sa"
 )
 
+const clientName = "SA"
+
+type config struct {
+	SA struct {
+		cmd.ServiceConfig
+		cmd.DBConfig
+
+		MaxConcurrentRPCServerRequests int64
+	}
+
+	cmd.StatsdConfig
+
+	cmd.SyslogConfig
+}
+
 func main() {
-	app := cmd.NewAppShell("boulder-sa", "Handles SQL operations")
-	app.Action = func(c cmd.Config, stats metrics.Statter, logger blog.Logger) {
-		saConf := c.SA
-		go cmd.DebugServer(saConf.DebugAddr)
-
-		dbURL, err := saConf.DBConfig.URL()
-		cmd.FailOnError(err, "Couldn't load DB URL")
-		dbMap, err := sa.NewDbMap(dbURL, saConf.DBConfig.MaxDBConns)
-		cmd.FailOnError(err, "Couldn't connect to SA database")
-		go sa.ReportDbConnCount(dbMap, metrics.NewStatsdScope(stats, "SA"))
-
-		sai, err := sa.NewSQLStorageAuthority(dbMap, clock.Default(), logger)
-		cmd.FailOnError(err, "Failed to create SA impl")
-
-		go cmd.ProfileCmd("SA", stats)
-
-		amqpConf := saConf.AMQP
-		sas, err := rpc.NewAmqpRPCServer(amqpConf, c.SA.MaxConcurrentRPCServerRequests, stats, logger)
-		cmd.FailOnError(err, "Unable to create SA RPC server")
-		err = rpc.NewStorageAuthorityServer(sas, sai)
-		cmd.FailOnError(err, "Unable to setup SA RPC server")
-
-		err = sas.Start(amqpConf)
-		cmd.FailOnError(err, "Unable to run SA RPC server")
+	configFile := flag.String("config", "", "File path to the configuration file for this service")
+	flag.Parse()
+	if *configFile == "" {
+		flag.Usage()
+		os.Exit(1)
 	}
 
-	app.Run()
+	var c config
+	err := cmd.ReadJSONFile(*configFile, &c)
+	cmd.FailOnError(err, "Reading JSON config file into config structure")
+
+	saConf := c.SA
+
+	go cmd.DebugServer(saConf.DebugAddr)
+
+	dbURL, err := saConf.DBConfig.URL()
+	cmd.FailOnError(err, "Couldn't load DB URL")
+
+	dbMap, err := sa.NewDbMap(dbURL, saConf.DBConfig.MaxDBConns)
+	cmd.FailOnError(err, "Couldn't connect to SA database")
+
+	stats, logger := cmd.StatsAndLogging(c.StatsdConfig, c.SyslogConfig)
```

Comment:
Let's move this section right below where we read the config, and right above the `DebugServer` command. The `saConf :=` line can go below that whole chunk.

The reasoning is that I'd like the bits that are common across many of the binaries to all be in one place so that it's more straightforward to factor them out into a separate function in the future if we choose to do that. Also that means it's more straightforward to check that each expected piece is there.